openhtf/util/validators.py

```diff
@@ -50,124 +50,140 @@ def MyPhase(test):
+  """Validator to verify a numeric value is within a range."""
+
+  def __init__(self, minimum, maximum, type=None):
+    if minimum is None and maximum is None:
+      raise ValueError('Must specify minimum, maximum, or both')
+    if (minimum is not None and maximum is not None
+        and isinstance(minimum, numbers.Number)
+        and isinstance(maximum, numbers.Number)
+        and minimum > maximum):
+      raise ValueError('Minimum cannot be greater than maximum')
+    self.minimum = minimum
+    self.maximum = maximum
+    self._type = type or (lambda x: x)
+
+  def with_args(self, **kwargs):
+    return type(self)(
+        minimum=util.format_string(self.minimum, kwargs),
+        maximum=util.format_string(self.maximum, kwargs),
+        type=self._type,
+    )
+
+  def __call__(self, value):
+    if value is None:
+      return False
+    import math
+    # Check for nan
+    if math.isnan(value):
+      return False
+    minimum = self._type(self.minimum)
+    maximum = self._type(self.maximum)
+    # Check for equal bounds first so we can use with non-numeric values.
+    if minimum == maximum and value != minimum:
+      return False
+    if minimum is not None and value < minimum:
+      return False
+    if maximum is not None and value > maximum:
+      return False
+    return True
+
+  def __str__(self):
+    assert self.minimum is not None or self.maximum is not None
+    if self.minimum is not None and self.maximum is not None:
+      if self.minimum == self.maximum:
+        return 'x == %s' % self.minimum
+      return '%s <= x <= %s' % (self.minimum, self.maximum)
+    if self.minimum is not None:
+      return '%s <= x' % self.minimum
+    if self.maximum is not None:
+      return 'x <= %s' % self.maximum
+
+  def __eq__(self, other):
+    return (isinstance(other, type(self)) and
+            self.minimum == other.minimum and self.maximum == other.maximum)
+
+  def __ne__(self, other):
+    return not self == other
+
+in_range = InRange  # pylint: disable=invalid-name
+register(in_range, name='in_range')
+
+
+@register
+def equals(self, value, type=None):
+  if isinstance(value, numbers.Number):
+    return InRange(minimum=value, maximum=value, type=type)
+  elif isinstance(value, basestring):
+    assert type is None or issubclass(type, basestring), (
+        'Cannot use a non-string type when matching a string')
+    return matches_regex(re.escape(value))
+  else:
+    return Equals(value, type=type)
+
+
+class Equals(object):
+  """Validator to verify an object is equal to the expected value."""
+
+  def __init__(self, expected, type=None):
+    self.expected = expected
+    self._type = type or (lambda x: x)
+
+  def __call__(self, value):
+    return value == self._type(self.expected)
+
+  def __str__(self):
+    return "'x' is equal to '%s'" % self.expected
+
+  def __eq__(self, other):
+    return isinstance(other, type(self)) and self.expected == other.expected
+
+
+class RegexMatcher(object):
```

Comment:
maybe postfix these helper classes with _, since the intention is for users to always use the lower_case validators.